felix/nftables/table.go

```diff
@@ -0,0 +1,1294 @@
+		}
+	}
+}
+
+// decrefChain decrements the refcount of the given chain; if the refcount transitions to 0,
+// marks the chain dirty so it will be cleaned up.
+func (t *nftablesTable) decrefChain(chainName string) {
+	t.logCxt.WithField("chainName", chainName).Debug("Decref chain")
+	if t.chainRefCounts[chainName] == 1 {
+		t.updateRateLimitedLog.WithField("chainName", chainName).Info("Chain no longer referenced, marking it for removal")
+		if chain := t.chainNameToChain[chainName]; chain != nil {
+			// Recursively decref chains that this chain refers to.  If
+			// chain == nil then the chain has probably already been deleted
+			// in which case we'll already have done the decrefs.
+			t.maybeDecrefReferredChains(chainName, chain.Rules)
+		}
+		delete(t.chainRefCounts, chainName)
+		t.dirtyChains.Add(chainName)
+		return
+	}
+
+	// Chain still referenced, just decrement.
+	t.chainRefCounts[chainName] -= 1
+}
+
+func (t *nftablesTable) loadDataplaneState() {
+	// Refresh the cache of feature data.
+	t.featureDetector.RefreshFeatures()
+
+	// Load the hashes from the dataplane.
+	t.logCxt.Debug("Loading current nftables state and checking it is correct.")
+	t.opReporter.RecordOperation(fmt.Sprintf("resync-%v-v%d", t.name, t.ipVersion))
+
+	t.lastReadTime = t.timeNow()
+
+	dataplaneHashes, dataplaneRules := t.getHashesAndRulesFromDataplane()
+
+	// Check that the rules we think we've programmed are still there and mark any inconsistent
+	// chains for refresh.
+	for chainName, expectedHashes := range t.chainToDataplaneHashes {
+		logCxt := t.logCxt.WithField("chainName", chainName)
+		if t.dirtyChains.Contains(chainName) || t.dirtyBaseChains.Contains(chainName) {
+			// Already an update pending for this chain; no point in flagging it as
+			// out-of-sync.
+			logCxt.Debug("Skipping known-dirty chain")
+			continue
+		}
+		dpHashes := dataplaneHashes[chainName]
+		if !t.ourChainsRegexp.MatchString(chainName) {
+			// Not one of our chains, or any Calico table's chains - so it may be one that we're inserting rules into.
+			insertedRules := t.chainToInsertedRules[chainName]
+			if len(insertedRules) == 0 {
+				// This chain shouldn't have any inserts, make sure that's the
+				// case.  This case also covers the case where a chain was removed,
+				// making dpHashes nil.
+				dataplaneHasInserts := false
+				for _, hash := range dpHashes {
+					if hash != "" {
+						dataplaneHasInserts = true
+						break
+					}
+				}
+				if dataplaneHasInserts {
+					logCxt.WithField("actualRuleIDs", dpHashes).Warn(
+						"Chain had unexpected inserts, marking for resync")
+					t.dirtyBaseChains.Add(chainName)
+				}
+				continue
+			}
+
+			// Re-calculate the expected rule insertions based on the current length
+			// of the chain (since other processes may have inserted/removed rules
+			// from the chain, throwing off the numbers).
+			expectedHashes, _, _ = t.expectedHashesForInsertAppendChain(
```

Comment:
We shouldn't ever hit this block in nftables mode, since the insert / append chains are chains that we own.